File: Java Solutions/542_01_Matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        int [][] newMatrix = new int[m][n];

        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    newMatrix[i][j] = 0;
                } else {
                    newMatrix[i][j] = n+m;
                }
            }
        }

        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++) {
                if (newMatrix[i][j] == 0) {
                    optimize_neighbours(i, j, newMatrix, m, n);
                }
            }
        }
        return newMatrix;
    }

    void optimize_neighbours(int i, int j, int[][] matrix, int m, int n) {
        int value = matrix[i][j];
        if (i > 0) {
            if (matrix[i-1][j] - 1 > value) {
                matrix[i-1][j] = value + 1;
                optimize_neighbours(i-1, j, matrix, m, n);
            }
        }
        if (i < m - 1 ) {
            if (matrix[i+1][j] - 1 > value) {
                matrix[i+1][j] = value + 1;
                optimize_neighbours(i+1, j, matrix, m, n);
            }
        }
        if (j > 0 ) {
            if (matrix[i][j-1] - 1 > value) {
                matrix[i][j-1] = value + 1;
                optimize_neighbours(i, j-1, matrix, m, n);
            }
        }
        if (j < n - 1 ) {
            if (matrix[i][j+1] - 1 > value) {
                matrix[i][j+1] = value + 1;
                optimize_neighbours(i, j+1, matrix, m, n);
            }
        }


    }

}

File: Python Solutions/542_01_Matrix.py
from typing import List

# Given an m x n binary matrix mat, return the distance
# of the nearest 0 for each cell.
# The distance between two adjacent cells is 1.

# Do something similar to BFS:
# Just optimize the box with every run.

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m = len(mat)
        n = len(mat[0])
        new_matrix = [[m + n for j in range(n)] for i in range(m)]

        # # Initialize 0 values first
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    new_matrix[i][j] = 0

        def optimize_neighbours(i, j):
            value = new_matrix[i][j]
            if i > 0 :
                if new_matrix[i-1][j] - 1 > value:
                    new_matrix[i-1][j] = value + 1
                    optimize_neighbours(i-1, j)
            if i < m - 1 :
                if new_matrix[i + 1][j] - 1 > value:
                    new_matrix[i + 1][j] = value + 1
                    optimize_neighbours(i + 1, j)
            if j > 0 :
                if new_matrix[i][j - 1] - 1 > value:
                    new_matrix[i][j - 1] = value + 1
                    optimize_neighbours(i, j - 1)
            if j < n - 1 :
                if new_matrix[i][j + 1] - 1 > value:
                    new_matrix[i][j + 1] = value + 1
                    optimize_neighbours(i, j + 1)

        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    optimize_neighbours(i,j)

        return new_matrix


a = Solution()
print(a.updateMatrix([[0,0,0],[0,1,0],[0,0,0]]))

# Notes: python is too slow to compute this. Leetcode time limit exceeded.
# See java solution.
